services/src/main/java/edu/unc/lib/dl/cdr/services/rest/DatastreamRestController.java

```diff
@@ -91,14 +107,35 @@ private void recordDownloadEvent(PID pid, String datastream, AccessGroupSet prin
     @RequestMapping("/thumb/{pid}")
     public void getThumbnailSmall(@PathVariable("pid") String pid,
             @RequestParam(value = "size", defaultValue = "small") String size, HttpServletRequest request,
-            HttpServletResponse response) {
-        // TODO implement retrieval of derivatives
+            HttpServletResponse response) throws IOException {
+
+        getThumbnail(pid, size, request, response);
     }
 
     @RequestMapping("/thumb/{pid}/{size}")
-    public void getThumbnail(@PathVariable("pid") String pid,
+    public void getThumbnail(@PathVariable("pid") String pidString,
             @PathVariable("size") String size, HttpServletRequest request,
-            HttpServletResponse response) {
-        // TODO implement retrieval of derivatives
+            HttpServletResponse response) throws IOException {
+
+        PID pid = PIDs.get(pidString);
+        AccessGroupSet principals = getAgentPrincipals().getPrincipals();
+        String thumbName = "thumbnail_" + size.toLowerCase().trim();
+
+        derivativeContentService.streamData(pid, thumbName, principals, false, response);
+    }
+
+    @ResponseStatus(value = HttpStatus.NOT_FOUND)
+    @ExceptionHandler({ResourceNotFoundException.class, NotFoundException.class, FileNotFoundException.class})
+    public void handleResourceNotFound() {
+    }
+
+    @ResponseStatus(value = HttpStatus.FORBIDDEN)
+    @ExceptionHandler(AccessRestrictionException.class)
+    public void handleInvalidRecordRequest() {
+    }
+
+    @ResponseStatus(value = HttpStatus.BAD_REQUEST)
+    @ExceptionHandler({ObjectTypeMismatchException.class, IllegalArgumentException.class})
+    public void handleBadRequestException() {
```

Comment:
i wonder whether just handleBadRequest() would be a clearer name for this method, since the exception(s) handled are not BadRequestExceptions? this would also match your usage for the previous two methods. also, are you planing at some point to have some logic inside these methods, or are they simply meant as placecholders?

Reply:
Fixed. The methods are empty for the time being, at some point we may want to fill them in with an error response, but right now they just serve the purpose of setting a response code and returning a blank body when their exception is thrown.